src/menu_events.cpp

```diff
@@ -266,92 +266,105 @@ void menu_handler::recruit(int side_num, const map_location& last_hex)
 	}
 
 	if(!unknown_units.empty()) {
-		auto unknown_ids = utils::format_conjunct_list("", unknown_units);
 		// TRANSLATORS: An error that should never happen, might happen when loading an old savegame. If there are
 		// any units that the player can recruit then their standard recruitment dialog will be shown after this
 		// error message, otherwise they'll get the "You have no units available to recruit." error after this one.
-		const auto message = VNGETTEXT("Error: there’s an unknown unit type on your recruit list: $unknown_ids",
+		gui2::show_transient_message("", VNGETTEXT(
+			"Error: there’s an unknown unit type on your recruit list: $unknown_ids",
 			"Error: there are several unknown unit types on your recruit list: $unknown_ids",
 			unknown_units.size(),
-			utils::string_map { { "unknown_ids", unknown_ids }});
-		gui2::show_transient_message("", message);
+			{{ "unknown_ids", utils::format_conjunct_list("", unknown_units) }}
+		));
 	}
 
 	if(recruit_list.empty()) {
 		gui2::show_transient_message("", _("You have no units available to recruit."));
 		return;
 	}
 
-	const auto& dlg = units_dialog::build_recruit_dialog(recruit_list, err_msgs_map, current_team);
-	dlg->set_selected_index(selected);
+	auto dlg = units_dialog::build_recruit_dialog(recruit_list, err_msgs_map, current_team);
+	dlg->set_selected_index(selected_index);
 	dlg->show();
-	const auto& type = recruit_list[dlg->get_selected_index()];
+
+	const unit_type* type = recruit_list[dlg->get_selected_index()];
 	last_recruit = type->id();
 
-	if((dlg->get_retval() == gui2::retval::OK) && dlg->is_selected()) {
-		map_location recruit_hex = last_hex;
-		do_recruit(type->id(), side_num, recruit_hex);
+	if(dlg->get_retval() == gui2::retval::OK && dlg->is_selected()) {
+		do_recruit(type->id(), side_num, last_hex);
 	}
 }
 
 void menu_handler::repeat_recruit(int side_num, const map_location& last_hex)
 {
-	const std::string& last_recruit = board().get_team(side_num).last_recruit();
-	if(!last_recruit.empty()) {
-		map_location recruit_hex = last_hex;
-		do_recruit(last_recruit, side_num, recruit_hex);
+	team& recruiter = board().get_team(side_num);
+	const std::string& last_recruit = recruiter.last_recruit();
+	if(last_recruit.empty()) return;
+
+	std::string error = unit_helper::check_recruit_purse(
+		unit_types.find(last_recruit)->cost(), recruiter.gold(), unit_helper::planned_gold_spent(side_num));
+
+	if(error.empty()) {
+		do_recruit(last_recruit, side_num, last_hex);
+	} else {
+		gui2::show_transient_message("", error);
 	}
 }
 
-bool menu_handler::do_recruit(const std::string& name, int side_num, map_location& loc)
+bool menu_handler::do_recruit(const std::string& type, int side_num, const map_location& target_hex)
 {
-	map_location recruited_from = map_location::null_location();
-	team& current_team = board().get_team(side_num);
-	const auto& res = unit_helper::recruit_message(name, loc, recruited_from, current_team);
+	if(auto wb = pc_.get_whiteboard()) {
```

Comment:
Not sure if the error check needs to be before or after the whiteboard codepath...